### PR TITLE
Feature/BDN-687: Changed auction cancel logic

### PR DESCRIPTION
**Pull request by AntonZarytski**
https://appodeal.atlassian.net/browse/BDN-687

### CODE REVIEW

**Comment by AntonZarytski** [comment]
* Adding `ensureActive()` solves the problem of not instantly canceling a job, because of which, 2 stats requests may be sent
* Changing `adUnitQueue.poll()` to `adUnitQueueue.peek()` before the request starts, and `adUnitQueueue.peek()` to `adUnitQueueue.poll()` after the request completes, ensures that the current requested AdUnit is not lost in the case of `AuctionTimeout `

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 1]
```diff
@@ -139,8 +142,9 @@ internal class AuctionImpl(
     }
 
     override fun cancel(onFailure: (AuctionInfo?, Throwable) -> Unit) {
+        logInfo(TAG, "Trying to cancel auction. Is active: ${job?.isActive}")
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt:160]: onFailure calls onDestroy
Снимок экрана 2024-07-18 в 17 37 29

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 2]
```diff
@@ -261,6 +258,7 @@ internal class AuctionImpl(
     }
 
     private fun clearData() {
+        resultsCollector.clearRoundResults()
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: ^ called in `resultsCollector.clear()`

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 3]
```diff
@@ -122,54 +120,46 @@ internal class AuctionImpl(
...
             }
         }
+        // Finish auction
+        state.value = AuctionState.Finished
+        clearData()
+    }
+
+    override fun cancel() {
+        logInfo(TAG, "Trying to cancel auction. Is active: ${job?.isActive}")
+        if (job?.isActive == true) {
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt:161]: `auctionStat.markAuctionCanceled()` is missing

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 4]
```diff
@@ -122,54 +120,46 @@ internal class AuctionImpl(
...
+            adTypeParam.activity.runOnUiThread {
+                onFailure(auctionInfo, cause)
+            }
+        } ?: {
+            adTypeParam.activity.runOnUiThread {
+                onFailure(null, cause)
             }
         }
+        // Finish auction
+        state.value = AuctionState.Finished
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt:155]: `state.value = AuctionState.Finished` for what?

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt (hunk 3 above)
Comment (da2gl) [reply bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt:161]: if use that all adunits marked as AUCTION_CANCELLED
[this logic ](https://github.com/bidon-io/bidon-sdk-android/commit/7bc69ec849fb12c477e8c4fc70b863a7761c9cec)

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/ExecuteAuctionUseCaseImpl.kt [hunk 5]
```diff
@@ -56,8 +57,7 @@ internal class ExecuteAuctionUseCaseImpl(
 
                 while (adUnitQueue.isNotEmpty()) {
 
-                    val adUnit = adUnitQueue.poll()
-
+                    val adUnit = adUnitQueue.peek()
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/ExecuteAuctionUseCaseImpl.kt:61]: if we use peek we break `if (adUnit.pricefloor < pricefloor)` logic

Comment (da2gl) [reply bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/ExecuteAuctionUseCaseImpl.kt:61]: BelowPricefloor / Lose

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/ExecuteAuctionUseCase.kt [hunk 6]
```diff
@@ -24,5 +24,5 @@ internal interface ExecuteAuctionUseCase {
         tokens: Map<String, TokenInfo>
     )
 
-    suspend fun cancel(resultsCollector: ResultsCollector)
+    fun cancel(resultsCollector: ResultsCollector)
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/ExecuteAuctionUseCase.kt]: not used anywhere

**Comment by da2gl** [comment]
CheckList:
Adtypes BANNER/INTER/REWARDED
-Init + destroy
-Create + destroy
-Create + Load + Destroy (до начала опроса)
-Create + Load + Destroy(в момент опроса)
-Create + Load + Destroy(после опроса)
-После загрузки
-В момент показа
Основная проблема -> на прошлой версии данный exception воспроизводился на
-Create + Load + Destroy(в момент опроса)
Сейчас все ок
+ перепроверил https://appodeal.atlassian.net/browse/BDN-663?focusedCommentId=99369

**Pull request merged**